middleware/basicauth/basicauth.go

```diff
@@ -64,6 +73,74 @@ func (a BasicAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 // file or directory paths.
 type Rule struct {
 	Username  string
-	Password  string
+	Password  func(string) bool
 	Resources []string
 }
+
+type PasswordMatcher func(pw string) bool
+
+var (
+	htpasswords   map[string]map[string]PasswordMatcher
+	htpasswordsMu sync.Mutex
+)
+
+func GetHtpasswdMatcher(filename, username string) (PasswordMatcher, error) {
+	filename, err := filepath.Abs(filename)
```

Comment:
Before this, is the filename relative to the site root or the pwd? It seems to me it's from the pwd, and I'm wondering if it would make more sense to be relative to site root. What do you think?